File: src/types/modelRegistry.ts
export type ModelProductGroup =
  | 'available_five'
  | 'blocked'
  | 'backlog'
  | 'unknown';

export type ModelUiExecutionState =
  | 'closed_available'
  | 'license_blocked'
  | 'backlog'
  | 'unknown';

export interface ModelRegistryEntry {
  model_id: string;
  display_name: string;
  category: string;
  status:
    | 'available'
    | 'installed'
    | 'not_connected'
    | 'planned'
    | 'disabled'
    | 'probed'
    | 'parked'
    | 'pending_probe'
    | 'pending_registry'
    | 'controlled_smoke_verified'
    | 'smoke_rerun_verified'
    | 'closed'
    | 'blocked_license'
    | 'backlog';
  status_reason?: string | null;
  description: string;
  supports_probe: boolean;
  supports_dry_run: boolean;
  supports_real_run: boolean;
  supports_structure_output: boolean;
  supports_sequence_output: boolean;
  supports_ranking: boolean;
  output_artifact_types: string[];
  adapter_id: string;
  safety_note: string;
  validation_policy: string;
  real_run_enabled: boolean;
  stage: string;
  notes?: string | null;
  // P32A display-only readiness fields (do not overwrite canonical stage)
  readiness_gate?: string | null;
  readiness_level?: string | null;
  execution_locked: boolean;
  blocker_code?: string | null;
  next_authorization?: string | null;
  last_verified_at?: string | null;
  evidence_ref?: string | null;
  // P33K product-group UI governance fields
  product_group: ModelProductGroup;
  ui_selectable: boolean;
  ui_execution_state: ModelUiExecutionState;
  activation_requirements: string;
  delivery_status: string;
  /** P33P: structured PPFlow evidence metadata (ppflow only). */
  evidence?: P33PFlowEvidence | null;
}

export interface P33PFlowEvidenceSha256 {
  success_report: string;
  execution_manifest: string;
  result_manifest: string;
  status: string;
}

/**
 * P33P: the only evidence identifiers the UI is allowed to render or download.
 * Unknown ids returned by the API MUST be filtered out before rendering.
 */
export type P33PFlowEvidenceId = 'success_report' | 'execution_manifest' | 'result_manifest' | 'status';

export interface P33PFlowEvidence {
  success_report: string;
  execution_manifest: string;
  result_manifest: string;
  status: string;
  artifact_dir: string;
  sha256: P33PFlowEvidenceSha256;
  exit_code: number;
  elapsed_seconds: number;
  sample_dirs: number;
  file_count: number;
  artifact_bytes: number;
  validation_status: string;
}

export interface P33PFlowEvidenceResponse {
  model_id: string;
  stage: string;
  execution_locked: boolean;
  real_run_enabled: boolean;
  validation_status: string;
  evidence: P33PFlowEvidence;
  downloadable_ids: P33PFlowEvidenceId[];
}

/**
 * P33Q: unified six-model evidence contract. Every registered model returns
 * this shape from GET /models/{id}/evidence. Unknown evidence ids MUST be
 * filtered against KnownEvidenceId before rendering or downloading.
 */
export type KnownEvidenceId =
  | 'success_report'
  | 'execution_manifest'
  | 'result_manifest'
  | 'status'
  | 'reasonix_review'
  | 'artifacts_index'
  | 'delivery_manifest'
  | 'p3b_report'
  | 'p3c_report';

export type ModelEvidenceAvailability = 'available' | 'partial' | 'unavailable' | 'pending';

export interface ModelEvidenceArtifact {
  id: string;
  filename: string;
  media_type: string;
  sha256: string;
  bytes: number | null;
}

/** Flexible numeric stats; keys vary per model (only truthfully-known fields). */
export type ModelEvidenceStats = Record<string, number>;

export interface ModelEvidenceResponse {
  model_id: string;
  stage: string;
  validation_status: string;
  execution_locked: boolean;
  real_run_enabled: boolean;
  supports_real_run: boolean;
  evidence_ref: string;
  availability: ModelEvidenceAvailability;
  stats: ModelEvidenceStats | null;
  downloadable_ids: string[];
  artifacts: ModelEvidenceArtifact[];
}

export interface ModelSafetyFlags {
  executed_model: boolean;
  generated_candidates: boolean;
  generated_structure: boolean;
  generated_msa: boolean;
  is_scientific_result: boolean;
  computational_prediction_only: boolean;
  validation_status: string;
}

export interface ModelsListResponse {
  models: ModelRegistryEntry[];
  scientific_boundary: string;
  available_models: number;
  blocked_models: number;
  backlog_models: number;
  available_model_ids: string[];
  blocked_model_ids: string[];
  backlog_model_ids: string[];
}

export interface ModelRegistryStatusEntry {
  model_id: string;
  display_name: string;
  status: string;
  status_reason?: string | null;
  stage: string;
  supports_probe: boolean;
  supports_dry_run: boolean;
  supports_real_run: boolean;
  notes?: string | null;
  // P33K product-group UI governance fields
  product_group: ModelProductGroup;
  ui_selectable: boolean;
  ui_execution_state: ModelUiExecutionState;
  activation_requirements: string;
  delivery_status: string;
}

export interface ModelRegistryStatusResponse {
  models: ModelRegistryStatusEntry[];
  parked_models: string[];
  pending_probe_models: string[];
  pending_registry_models: string[];
  scientific_boundary: string;
  count_total: number;
  count_parked: number;
  count_pending_probe: number;
  count_pending_registry: number;
}

export interface ModelDetailResponse {
  model: ModelRegistryEntry;
  safety_flags: ModelSafetyFlags;
  scientific_boundary: string;
}

export interface ModelProbeResult {
  model_id: string;
  display_name: string;
  status: string;
  message: string;
  probe_time: string;
  adapter_id: string;
  safety_flags: ModelSafetyFlags;
  detail: Record<string, unknown>;
  scientific_boundary: string;
}

export interface ModelDryRunPayload {
  target_sequence: string;
  peptide_length: number;
  model_name: string;
  max_recycles: number;
  num_iterations: number;
  dry_run: boolean;
  // PepMLM / PepPrCLIP / PepGLAD specific optional fields
  num_candidates?: number;
  max_length?: number;
  top_k?: number;
  device?: 'auto' | 'cpu' | 'cuda';
  seed?: number | null;
  candidate_peptides?: string | string[] | null;
  // PepGLAD structure-conditioned inputs
  target_pdb_path?: string | null;
  pocket_residues?: string[] | null;
}

export interface ModelDryRunResult {
  model_id: string;
  display_name: string;
  status: string;
  message: string;
  run_id?: string | null;
  artifacts: Record<string, string | null>;
  expected_artifacts?: Record<string, string | null>;
  command_preview?: string[] | null;
  env_preview: Record<string, string>;
  environment_summary?: Record<string, unknown>;
  blocked_reasons?: string[];
  safety_flags: ModelSafetyFlags;
  validation_status: string;
  scientific_boundary: string;
}

export interface ModelArtifactItem {
  name: string;
  path: string;
  artifact_type: string;
  exists: boolean;
  size_bytes: number;
  download_url?: string | null;
}

export interface ModelArtifactsResponse {
  model_id: string;
  job_id: string;
  status: string;
  validation_status: string;
  safety_note: string;
  artifacts: ModelArtifactItem[];
}

export interface ModelJobStatus {
  job_id: string;
  model_id: string;
  status: string;
  progress?: number | null;
  message?: string | null;
  error_message?: string | null;
  error_json?: Record<string, unknown> | null;
  output_json?: Record<string, unknown> | null;
  started_at?: string | null;
  finished_at?: string | null;
  validation_status: string;
}

export interface PepmlmCandidate {
  candidate_id: string;
  sequence: string;
  length?: number | string | null;
  peptide_length?: number | string | null;
  source_model?: string | null;
  job_id?: string | null;
  validation_status?: string | null;
  safety_note?: string | null;
  pseudo_perplexity?: number | null;
  [key: string]: unknown;
}

export interface ModelSubmitResponse {
  job_id: string;
  status: string;
  message?: string | null;
  run_id?: string | null;
  artifacts?: ModelArtifactItem[];
  validation_status?: string;
  safety_flags?: ModelSafetyFlags;
}


export interface TargetDesignWorkflowPayload {
  target_sequence: string;
  generator_model: string;
  ranker_model: string;
  structure_model: string;
  num_candidates: number;
  peptide_length: number;
  top_k: number;
  dry_run: boolean;
}

export interface WorkflowStep {
  step_id: string;
  name: string;
  model_id: string;
  status: string;
  message: string;
  blocked_reason?: string | null;
  command_preview?: string[] | null;
  env_preview: Record<string, string>;
  artifacts: Record<string, string | null>;
  expected_artifacts: Record<string, string | null>;
}

export interface WorkflowArtifactReference {
  label: string;
  model_id: string;
  job_id?: string | null;
  artifact_name: string;
  artifact_path?: string | null;
  status: 'Exists' | 'Expected' | 'Blocked' | string;
  reason?: string | null;
  validation_status: string;
  download_url?: string | null;
}

export interface TargetDesignWorkflowResult {
  workflow_id: string;
  workflow_type: string;
  status: string;
  message: string;
  steps: WorkflowStep[];
  expected_artifacts: Record<string, string | null>;
  safety_flags: ModelSafetyFlags;
  validation_status: string;
  scientific_boundary: string;
  blocked_reasons: string[];
  prior_artifacts_referenced: Record<string, string | null>;
  artifact_references: WorkflowArtifactReference[];
}

File: reports/p33l_six_model_real_run/server_snapshot/model_registry.py
"""Schemas for the unified Model Registry and Adapter Interface (P4A).

All scientific outputs carry validation_status=NOT_EXPERIMENTALLY_VALIDATED
and safety_flags marking them as computational predictions only.
"""

from __future__ import annotations

from typing import Any, Optional

from pydantic import ConfigDict, Field

from app.schemas.project import SchemaBase


class ModelSafetyFlags(SchemaBase):
    """Safety flags attached to every model adapter operation."""

    executed_model: bool = Field(default=False)
    generated_candidates: bool = Field(default=False)
    generated_structure: bool = Field(default=False)
    generated_msa: bool = Field(default=False)
    is_scientific_result: bool = Field(default=False)
    runs_model: bool = Field(default=False)
    generates_candidates: bool = Field(default=False)
    experimental_validation: bool = Field(default=False)
    computational_prediction_only: bool = Field(default=True)
    validation_status: str = Field(default="NOT_EXPERIMENTALLY_VALIDATED")


class ModelRegistryEntry(SchemaBase):
    """Standardized model metadata in the unified registry."""

    model_config = ConfigDict(from_attributes=True)

    model_id: str
    display_name: str
    category: str
    status: str = Field(
        ...,
        description=(
            "available | not_connected | planned | disabled | probed | "
            "parked | pending_probe | pending_registry"
        ),
    )
    status_reason: Optional[str] = Field(
        default=None,
        description="Short machine-readable reason for the current status (e.g. parked_due_to_cert_or_weight_access_issue).",
    )
    description: str
    supports_probe: bool = Field(default=False)
    supports_dry_run: bool = Field(default=False)
    supports_real_run: bool = Field(default=False)
    supports_structure_output: bool = Field(default=False)
    supports_sequence_output: bool = Field(default=False)
    supports_ranking: bool = Field(default=False)
    output_artifact_types: list[str] = Field(default_factory=list)
    adapter_id: str
    safety_note: str = Field(default="Computational prediction only; NOT_EXPERIMENTALLY_VALIDATED.")
    validation_policy: str = Field(
        default=(
            "All outputs are computational predictions. "
            "They are NOT_EXPERIMENTALLY_VALIDATED and must not be interpreted as "
            "experimentally confirmed binding peptides, Kd, MIC, MM-GBSA, ipTM, pLDDT, RMSD, or RMSF."
        )
    )
    real_run_enabled: bool = Field(default=False)
    stage: str = Field(default="P4A")
    notes: Optional[str] = None
    validation_status: str = Field(default="NOT_EXPERIMENTALLY_VALIDATED")
    # P32A display-only readiness fields (do not overwrite canonical stage)
    readiness_gate: Optional[str] = Field(default=None)
    readiness_level: Optional[str] = Field(default=None)
    execution_locked: bool = Field(default=True)
    blocker_code: Optional[str] = Field(default=None)
    next_authorization: Optional[str] = Field(default=None)
    last_verified_at: Optional[str] = Field(default=None)
    evidence_ref: Optional[str] = Field(default=None)
    # P33K product-group UI governance fields
    product_group: str = Field(
        default="unknown",
        description="available_six | reserved_placeholder | excluded",
    )
    ui_selectable: bool = Field(default=False)
    ui_execution_state: str = Field(
        default="unknown",
        description="probe_dry_run_available | locked_placeholder | excluded",
    )
    activation_requirements: str = Field(default="explicit authorization required")
    delivery_status: str = Field(default="unknown")


class ModelsListResponse(SchemaBase):
    """Response for GET /api/v1/models."""

    models: list[ModelRegistryEntry]
    scientific_boundary: str


class ModelRegistryStatusEntry(SchemaBase):
    """Compact status entry for the 8-model track dashboard."""

    model_id: str
    display_name: str
    status: str
    status_reason: Optional[str] = None
    stage: str
    supports_probe: bool = Field(default=False)
    supports_dry_run: bool = Field(default=False)
    supports_real_run: bool = Field(default=False)
    notes: Optional[str] = None
    # P33K product-group UI governance fields
    product_group: str = Field(
        default="unknown",
        description="available_six | reserved_placeholder | excluded",
    )
    ui_selectable: bool = Field(default=False)
    ui_execution_state: str = Field(
        default="unknown",
        description="probe_dry_run_available | locked_placeholder | excluded",
    )
    activation_requirements: str = Field(default="explicit authorization required")
    delivery_status: str = Field(default="unknown")


class ModelRegistryStatusResponse(SchemaBase):
    """Response for GET /api/v1/model-registry/status."""

    models: list[ModelRegistryStatusEntry]
    parked_models: list[str] = Field(default_factory=list)
    pending_probe_models: list[str] = Field(default_factory=list)
    pending_registry_models: list[str] = Field(default_factory=list)
    scientific_boundary: str
    count_total: int
    count_parked: int
    count_pending_probe: int
    count_pending_registry: int


class ModelDetailResponse(SchemaBase):
    """Response for GET /api/v1/models/{model_id}."""

    model: ModelRegistryEntry
    safety_flags: ModelSafetyFlags
    scientific_boundary: str


class ModelProbeResult(SchemaBase):
    """Response for GET /api/v1/models/{model_id}/probe."""

    model_id: str
    display_name: str
    status: str
    message: str
    probe_time: str
    adapter_id: str
    safety_flags: ModelSafetyFlags
    detail: dict[str, Any] = Field(default_factory=dict)
    scientific_boundary: str
    # P1 skeleton blocked fields (optional, populated for pending_registry adapters)
    available: Optional[bool] = Field(default=None)
    actionable: Optional[bool] = Field(default=None)
    probe_status: Optional[str] = Field(default=None)
    reason: Optional[str] = Field(default=None)
    stage: Optional[str] = Field(default=None)
    runs_model: Optional[bool] = Field(default=None)
    generates_candidates: Optional[bool] = Field(default=None)
    experimental_validation: Optional[bool] = Field(default=None)
    is_scientific_result: Optional[bool] = Field(default=None)
    computational_prediction_only: Optional[bool] = Field(default=None)

    real_run_status: Optional[str] = Field(default=None)
    real_run_blocked_reason: Optional[str] = Field(default=None)
    generates_pdb: Optional[bool] = Field(default=None)
    creates_job: Optional[bool] = Field(default=None)
    writes_artifacts: Optional[bool] = Field(default=None)
    runs_subprocess: Optional[bool] = Field(default=None)
    would_use_checkpoint: Optional[str] = Field(default=None)
    would_use_env: Optional[str] = Field(default=None)
    mode: Optional[str] = Field(default=None)
    validation_status: Optional[str] = Field(default=None)


class ModelDryRunPayload(SchemaBase):
    """Request body for POST /api/v1/models/{model_id}/dry-run."""

    target_sequence: Optional[str] = Field(default=None, min_length=1)
    peptide_length: int = Field(default=10, ge=1, le=100)
    peptide_sequence: Optional[str] = Field(default=None)
    model_name: str = Field(default="model_1_ptm")
    max_recycles: int = Field(default=1, ge=1, le=10)
    num_iterations: int = Field(default=1, ge=1, le=100)
    dry_run: bool = Field(default=True)
    # PepMLM / PepPrCLIP / PepGLAD specific optional fields
    num_candidates: int = Field(default=5, ge=1, le=100)
    max_length: int = Field(default=10, ge=1, le=200)
    top_k: int = Field(default=3, ge=1, le=100)
    device: str = Field(default="auto")
    seed: Optional[int] = Field(default=None, ge=0)
    candidate_peptides: Optional[str | list[str]] = Field(default=None)
    # PepGLAD structure-conditioned inputs
    target_pdb_path: Optional[str] = Field(default=None)
    target_chain: str = Field(default="A", min_length=1, max_length=8)
    pocket_residues: Optional[list[str]] = Field(default=None)


class ModelDryRunResult(SchemaBase):
    """Response for POST /api/v1/models/{model_id}/dry-run."""

    model_id: str
    display_name: str
    status: str
    message: str
    run_id: Optional[str] = None
    artifacts: dict[str, Optional[str]] = Field(default_factory=dict)
    expected_artifacts: dict[str, Optional[str]] = Field(default_factory=dict)
    expected_inputs: dict[str, Any] = Field(default_factory=dict)
    expected_outputs: dict[str, Any] = Field(default_factory=dict)
    artifact_plan: dict[str, Any] = Field(default_factory=dict)
    command_preview: Optional[list[str]] = None
    env_preview: dict[str, str] = Field(default_factory=dict)
    environment_summary: dict[str, Any] = Field(default_factory=dict)
    blocked_reasons: list[str] = Field(default_factory=list)
    safety_flags: ModelSafetyFlags
    validation_status: str = Field(default="NOT_EXPERIMENTALLY_VALIDATED")
    scientific_boundary: str
    # P1 skeleton blocked fields (optional, populated for pending_registry adapters)
    available: Optional[bool] = Field(default=None)
    actionable: Optional[bool] = Field(default=None)
    dry_run_status: Optional[str] = Field(default=None)
    reason: Optional[str] = Field(default=None)
    stage: Optional[str] = Field(default=None)
    runs_model: Optional[bool] = Field(default=None)
    generates_candidates: Optional[bool] = Field(default=None)
    experimental_validation: Optional[bool] = Field(default=None)
    is_scientific_result: Optional[bool] = Field(default=None)
    computational_prediction_only: Optional[bool] = Field(default=None)

    real_run_status: Optional[str] = Field(default=None)
    real_run_blocked_reason: Optional[str] = Field(default=None)
    generates_pdb: Optional[bool] = Field(default=None)
    creates_job: Optional[bool] = Field(default=None)
    writes_artifacts: Optional[bool] = Field(default=None)
    runs_subprocess: Optional[bool] = Field(default=None)
    would_use_checkpoint: Optional[str] = Field(default=None)
    would_use_env: Optional[str] = Field(default=None)
    mode: Optional[str] = Field(default=None)



class TargetDesignWorkflowPayload(SchemaBase):
    """Request body for POST /api/v1/workflows/target-design/dry-run."""

    target_sequence: str = Field(..., min_length=1)
    generator_model: str = Field(default="pepmlm")
    ranker_model: str = Field(default="pepprclip")
    structure_model: str = Field(default="evobind2")
    num_candidates: int = Field(default=3, ge=1, le=100)
    peptide_length: int = Field(default=12, ge=1, le=200)
    top_k: int = Field(default=3, ge=1, le=100)
    dry_run: bool = Field(default=True)


class WorkflowStep(SchemaBase):
    """Single step in a target-design workflow dry-run plan."""

    step_id: str
    name: str
    model_id: str
    status: str
    message: str
    blocked_reason: Optional[str] = Field(default=None)
    command_preview: Optional[list[str]] = Field(default=None)
    env_preview: dict[str, str] = Field(default_factory=dict)
    artifacts: dict[str, Optional[str]] = Field(default_factory=dict)
    expected_artifacts: dict[str, Optional[str]] = Field(default_factory=dict)


class WorkflowArtifactReference(SchemaBase):
    """Existing or planned artifact in the target-design workflow lineage."""

    label: str
    model_id: str
    job_id: Optional[str] = None
    artifact_name: str
    artifact_path: Optional[str] = None
    status: str
    reason: Optional[str] = None
    validation_status: str = Field(default="NOT_EXPERIMENTALLY_VALIDATED")
    download_url: Optional[str] = None


class TargetDesignWorkflowResult(SchemaBase):
    """Response for POST /api/v1/workflows/target-design/dry-run."""

    workflow_id: str
    workflow_type: str = Field(default="target-design")
    status: str
    message: str
    steps: list[WorkflowStep]
    expected_artifacts: dict[str, Optional[str]] = Field(default_factory=dict)
    safety_flags: ModelSafetyFlags
    validation_status: str = Field(default="NOT_EXPERIMENTALLY_VALIDATED")
    scientific_boundary: str
    # P1 skeleton blocked fields (optional, populated for pending_registry adapters)
    available: Optional[bool] = Field(default=None)
    actionable: Optional[bool] = Field(default=None)
    dry_run_status: Optional[str] = Field(default=None)
    reason: Optional[str] = Field(default=None)
    stage: Optional[str] = Field(default=None)
    runs_model: Optional[bool] = Field(default=None)
    generates_candidates: Optional[bool] = Field(default=None)
    experimental_validation: Optional[bool] = Field(default=None)
    blocked_reasons: list[str] = Field(default_factory=list)
    prior_artifacts_referenced: dict[str, Optional[str]] = Field(default_factory=dict)
    artifact_references: list[WorkflowArtifactReference] = Field(default_factory=list)


class ModelArtifactItem(SchemaBase):
    """Single artifact metadata returned by the unified model API."""

    name: str
    path: str = Field(..., description="Internal relative path under the artifact root.")
    artifact_type: str = Field(default="other")
    exists: bool
    size_bytes: int = Field(default=0)
    download_url: Optional[str] = None


class ModelArtifactsResponse(SchemaBase):
    """Response for GET /api/v1/models/{model_id}/jobs/{job_id}/artifacts."""

    model_id: str
    job_id: str
    status: str
    validation_status: str = Field(default="NOT_EXPERIMENTALLY_VALIDATED")
    safety_note: str = Field(
        default=(
            "These are computational prediction artifacts only. "
            "They are NOT_EXPERIMENTALLY_VALIDATED and must not be interpreted as "
            "experimentally confirmed binding peptides."
        )
    )
    artifacts: list[ModelArtifactItem] = Field(default_factory=list)
